File: Data Structures & Algorithms/validate-parentheses/submission-5.py
class Solution:
    def isValid(self, s: str) -> bool:
        stack = []
        for c in s:
            if c == ')':
                if stack and stack.pop() != '(':
                    return False
            elif c == ']':
                if stack and stack.pop() != '[':
                    return False
            elif c == '}':
                if stack and stack.pop() != '{':
                    return False
            else :
                stack.append(c)

        if stack:
            return False
        return True

File: Data Structures & Algorithms/valid-parenthesis-string/submission-5.cpp
class Solution {
public:
    bool checkValidString(string s) {
        stack<int> left, star;
        for (int i{}; i < s.size(); ++i) {
            char c{s[i]};
            if (c == '(') {
                left.push(i);
            }
            else if (c == '*') {
                star.push(i);
            }
            else {
                if (!left.size() && !star.size()) return false;
                if (left.size()) left.pop();
                else star.pop();
            }
        }
        while (left.size() && star.size()) {
            if (left.top() > star.top()) return false;
            left.pop();
            star.pop();
        }
        return left.size() == 0;
    }
};
